### DISCUSSION TITLE
How to Dynamically Create Tables With Column Names and Constraints From Dictionary Using ORM's Declarative Base?

### DISCUSSION THREAD

**Discussion by fleetofthought** [sqlalchemy/sqlalchemy#7289] [Usage Questions] [answered]
Using SQLAlchemy 1.4, Postgres 3, Python 3.7

My question is regarding creating classes dynamically rather than relying on the contents of `models.py`. I have a
json file with schema for many tables. Number of columns, column names, column constraints differ from table to table and are not known in advance.

The JSON gets parsed and its results are mapped to the ORM Postgres dialect (ex: {'column_name1': 'bigint'} becomes 'column_name1 = Column(BigInt)'). This creates a dictionary that contains a table name, column names, and column constraints. Since all the tables get passed the Augmented Base they automatically receive a PK id field.

I then pass this dictionary to a `create_class` function that uses this data to create the tables dynamically and commit these new tables to the database.

The challenge is that when I run the code the tables do get created but only with a single column - that of the PK id which it received automatically. All the other columns are ignored. 

I suspect I am creating this error in the way that I am invoking the Session or Base or in the way I am passing the column constraints. I'm not sure how to indicate to the ORM that I am passing in Column and Constraint objects.

I have tried changing things like:
- the way the classes are created - passing in a Column object instead of a Column string
ex: `constraint_dict[k] = f'= Column({v})'` VS `constraint_dict[k] = f'= {Column}({v})'`

- changing the way the column constraints are collected

- calling `Base` and `create` in different ways. I try to show these variations in the commented out lines in `create_class` below.

I can't work out which bits of interaction are creating this error. Any help is much appreciated!

Here is the code:

**base.py**
```py
from sqlalchemy import Column, create_engine, Integer, MetaData
from sqlalchemy.orm import declared_attr, declarative_base, scoped_session, sessionmaker

engine = create_engine('postgresql://user:pass@localhost:5432/dev', echo=True)

db_session = scoped_session(
    sessionmaker(bind=engine,  autocommit=False, autoflush=False)
)
# Augment the base class by using the cls argument of the declarative_base() function so all classes derived
# from Base will have a table name derived from the class name and an id primary key column.
class Base:
    @declared_attr
    def __tablename__(cls):
        return cls.__name__.lower()

    id = Column(Integer, primary_key=True)

metadata_obj = MetaData(schema='collect')
Base = declarative_base(cls=Base, metadata=metadata_obj)
```

**models.py**

```py
from base import Base
from sqlalchemy import Column, DateTime, Integer, Text
from sqlalchemy.dialects.postgresql import UUID
import uuid


class NumLimit(Base):

    org = Column(UUID(as_uuid=True), default=uuid.uuid4, unique=True)
    limits = Column(Integer)
    limits_rate = Column(Integer)
    rate_use = Column(Integer)

    def __init__(self, org, limits, allowance_rate, usage, last_usage):
        super().__init__()
        self.org = org
        self.limits = limits
        self.limits_rate = limits_rate
        self.rate_use = rate_use
```

**create_tables.py** (I know this one is messy! Just trying to show all the variations attempted ...)
```py
def convert_snake_to_camel(name):
    return ''.join(x.capitalize() or '_' for x in name.split('_'))


def create_class(table_data):
    constraint_dict = {'__tablename__': 'TableClass'}
    table_class_name = ''
    column_dict = {}

    for k, v in table_data.items():

        # Retrieve table, alter the case, store it for later use
        if 'table' in k:
            constraint_dict['__tablename__'] = v
            table_class_name += convert_snake_to_camel(v)

        # Retrieve the rest of the values which are the column names and constraints, ex: 'org = Column(UUID(as_uuid=True), default=uuid.uuid4, unique=True)'
        else:
            constraint_dict[k] = f'= Column({v})'
            column_dict[k] = v

    # When type is called with 3 arguments it produces a new class object, so we use it here to create the table Class
    table_cls = type(table_class_name, (Base,), constraint_dict)

    # Call ORM's 'Table' on the Class
    # table_class = Table(table_cls) # Error "TypeError: Table() takes at least two positional-only arguments 'name' and 'metadata'"

    # db_session.add(table_cls) # Error "sqlalchemy.orm.exc.UnmappedInstanceError: Class 
  'sqlalchemy.orm.decl_api.DeclarativeMeta' is not mapped; was a class (__main__.Metadata) supplied where an instance was required?"

    # table_class = Table(
    #    table_class_name,
    #    Base.metadata,
    #    constraint_dict)   # Error "sqlalchemy.orm.exc.UnmappedInstanceError: Class 'sqlalchemy.orm.decl_api.DeclarativeMeta' is not mapped; was a class (__main__.Metadata) supplied where an instance was required?"

    # table_class = Table(
    #    table_class_name,
    #    Base.metadata,
    #    column_dict)  
    #    table_class.create(bind=engine, checkfirst=True) # sqlalchemy.exc.ArgumentError: 'SchemaItem' object, such as a 'Column' or a 'Constraint' expected, got {'limits': 'Integer'}
 
    # table_class = Table(
    #    table_class_name,
    #    Base.metadata,
    #    **column_dict) # TypeError: Additional arguments should be named <dialectname>_<argument>, got 'limits'

    # Base.metadata.create_all(bind=engine, checkfirst=True)
    # table_class.create(bind=engine, checkfirst=True)
    
    new_row_vals = table_cls(**column_dict)
    db_session.add(new_row_vals)  # sqlalchemy.exc.ArgumentError: 'SchemaItem' object, such as a 'Column' or a 'Constraint' expected, got {'limits': 'Integer'}

    db_session.commit()
    db_session.close()
```

**Comment by CaselIT** [comment]
Hi,

In the example you provided it's missing both the json and/or the data that you pass to `create_class`.
I would suggest looking at what `constraint_dict` contains and going from there.

From a quick look you seem to be adding strings to it `constraint_dict[k] = f'= Column({v})'`, while you probably want to add column instances.

**Reply by fleetofthought** [reply]
Thanks for the response! I've updated my question with more detail below in zzzeek's response.

**Comment by zzzeek** [comment]
hi -

I can show you how to dynamically create a declarative mapped class but it isn't exactly in terms of your example above, because there's things going on there I don't follow.  Mostly that it seems to be calling upon "Table" quite a lot, which when you use declarative mappings you don't have to, because [declarative creates a Table object for you](https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#orm-declarative-table).    Also I don't know why you would be referring to a Session and a database connection within the `create_class` function, which seems like it just wants to create a new declarative class, not work with any actual data.   So maybe I need to show you the scope from "create a new declarative class" to "insert some data".

since I don't know exactly what your JSON format is, I'll just do one here that is intuitive for the task at hand:

```py


{
    "clsname": "MyClass",
    "tablename": "my_table",
    "columns": [
          {"name": "col1", "type": "integer", "is_pk": True},
          {"name": "col2", "type": "string"},
          {"name": "col3", "type": "integer"},
    ]
}

```

then let's say somewhere else you have some data like this:

```py

data = [
    {"col1": 1, "col2": "the first object", "col3": 12},
    {"col1": 2, "col2": "the second object", "col3": 9},
    {"col1": 3, "col2": "the third object", "col3": 15},
]
```

that's the input.  ready?   this is the most barebones approach that makes no attempt to assert that the incoming data has the required fields or anything like that, here's the function by itself:

```py

_type_lookup = {"integer": Integer, "string": String}

def mapping_for_json(json_cls_schema):

    clsdict = {"__tablename__": json_cls_schema["tablename"]}

    clsdict.update(
        {
            rec["name"]: Column(
                _type_lookup[rec["type"]], primary_key=rec.get("is_pk", False)
            )
            for rec in json_cls_schema["columns"]
        }
    )
    return type(json_cls_schema["clsname"], (Base,), clsdict)
```


that's it!  so below is a POC you can run with the above data run into a SQLite database.    Hopefully you can use the general idea here and adapt it to the specific thing you are doing:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


_type_lookup = {"integer": Integer, "string": String}


def mapping_for_json(json_cls_schema):

    clsdict = {"__tablename__": json_cls_schema["tablename"]}

    clsdict.update(
        {
            rec["name"]: Column(
                _type_lookup[rec["type"]], primary_key=rec.get("is_pk", False)
            )
            for rec in json_cls_schema["columns"]
        }
    )
    return type(json_cls_schema["clsname"], (Base,), clsdict)


if __name__ == "__main__":

    json_cls_schema = {
        "clsname": "MyClass",
        "tablename": "my_table",
        "columns": [
            {"name": "col1", "type": "integer", "is_pk": True},
            {"name": "col2", "type": "string"},
            {"name": "col3", "type": "integer"},
        ],
    }

    data = [
        {"col1": 1, "col2": "the first object", "col3": 12},
        {"col1": 2, "col2": "the second object", "col3": 9},
        {"col1": 3, "col2": "the third object", "col3": 15},
    ]

    MyClass = mapping_for_json(json_cls_schema)

    e = create_engine("sqlite://", echo=True)
    Base.metadata.create_all(e)

    with Session(e) as sess:
        sess.add_all(MyClass(**rec) for rec in data)
        sess.commit()

```

output:

```
2021-11-04 09:05:48,949 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-11-04 09:05:48,949 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("my_table")
2021-11-04 09:05:48,949 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-11-04 09:05:48,949 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("my_table")
2021-11-04 09:05:48,949 INFO sqlalchemy.engine.Engine [raw sql] ()
2021-11-04 09:05:48,950 INFO sqlalchemy.engine.Engine 
CREATE TABLE my_table (
	col1 INTEGER NOT NULL, 
	col2 VARCHAR, 
	col3 INTEGER, 
	PRIMARY KEY (col1)
)


2021-11-04 09:05:48,950 INFO sqlalchemy.engine.Engine [no key 0.00007s] ()
2021-11-04 09:05:48,950 INFO sqlalchemy.engine.Engine COMMIT
2021-11-04 09:05:48,951 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-11-04 09:05:48,952 INFO sqlalchemy.engine.Engine INSERT INTO my_table (col1, col2, col3) VALUES (?, ?, ?)
2021-11-04 09:05:48,952 INFO sqlalchemy.engine.Engine [generated in 0.00012s] ((1, 'the first object', 12), (2, 'the second object', 9), (3, 'the third object', 15))
2021-11-04 09:05:48,953 INFO sqlalchemy.engine.Engine COMMIT
```

**Reply by fleetofthought** [reply]
Thanks for the response!

There's still something here that I'm not quite grokking.

Breaking it down:
At this point, adding in data to a table isn't required. I'm just focusing on:
1. Parsing the data from the JSON schema to derive our table data.
2. Creating a Table Class based on this table data.
3. Committing this empty table, with all its columns and constraints, to the database.

Actually adding in data will be a secondary step :)

So let's assume we are working with two tables: `User` and `Terms`.
Once we're done with step 1, we are left with `table_dict` which holds this information for each table:

```
{
    'tablename': 'user',
    'columns': [
        {'new_col': 'email', 'type': ['Text', 'nullable=False', 'unique=True']},
        {'new_col': 'name', 'type': ['Text']},
        {'new_col': 'org', 'type': ['Text', 'nullable=False', index=True']},
        {'new_col': 'last_seen', 'type': ['DateTime(timezone=True)']}
    ]
}

{
    'tablename': 'terms',
    'columns': [
        {'new_col': 'owner', 'type': ['UUID(as_uuid=True), default=uuid.uuid4, unique=True', 'nullable=False', 'index=True']},
        {'new_col': 'received', 'type': ['Boolean', 'nullable=False']},
        {'new_col': 'date_received', 'type': ['DateTime(timezone=True)', 'nullable=False']}
    ]
}
```
We pass this dict to `def create_class(table_dict)` and now we are at step 2, creating the TableClass.

Here we will add to table_dict a field like `'__tablename__': 'terms'` and create a class name `Terms` to conform to expected input style, and we will call Type on it before opening the Sessions.

```
def create_class(table_dict):

    table_dict['__tablename__'] = table_dict['tablename']
    table_dict['class_name'] = convert_snake_to_camel(table_dict['tablename'])
    TableClass = type(table_dict['class_name'], (Base,), table_dict)
```

An entry in `table_dict` will now look like:
```
{
    'tablename': 'terms',
    'columns': [
        {'new_col': 'owner', 'type': ['UUID(as_uuid=True), default=uuid.uuid4, unique=True', 'nullable=False', 'index=True']},
        {'new_col': 'received', 'type': ['Boolean', 'nullable=False']},
        {'new_col': 'date_received', 'type': ['DateTime(timezone=True)', 'nullable=False']},
    '__tablename__': 'terms',
    'class_name': 'Terms'
    ]
}
```
Now we are at step 3!
We want to commit this empty table, with all its columns and constraints, to the database.

We call our engine and Scoped session from base.py
For reference here is that code:
```
engine = create_engine('postgresql://user:pass@localhost:5432/dev', echo=True)
db_session = scoped_session( sessionmaker(bind=engine,
                                          autocommit=False,
                                          autoflush=False) )
class Base:
    @declared_attr
    def __tablename__(cls):
        return cls.__name__.lower()
    id = Column(Integer, primary_key=True)

metadata_vals = MetaData(schema='collect')
# Instantiate a Base class for our classes definitions
Base = declarative_base(cls=Base, metadata=metadata_vals)
```

So after we call our function and have our object `TableClass` we start the engine:

```
Base.metadata.create_all(bind=engine)
with db_session() as sess:
     sess.add_all(TableClass(**rec) for rec in TableClass)
     sess.commit()

```
This errors with `TypeError: 'DeclarativeMeta' object is not iterable` which makes sense.

If I pass in the dict as a whole instead of the TableClass object it errors `TypeError: DeclarativeMeta object argument after ** must be a mapping, not str` 
This makes sense given that we've ignored the TableClass object and are just dealing with a dict.

If I bypass `sess.add_all` (because we're not really adding instances, we're just trying to create the table) and just call `Base.metadata.create_all(bind=engine)`, the code doesn't do anything (it no longer even commits the default values such as PK id derived from our Augmented Base.

It seems like I'm still making the same error as in my initial question! Have you any thoughts on how to read the table dictionary correctly into the Class correctly into the database? Thank you!!

**Reply by zzzeek** [reply]
> ```
> def create_class(table_dict):
> 
>     table_dict['__tablename__'] = table_dict['tablename']
>     table_dict['class_name'] = convert_snake_to_camel(table_dict['tablename'])
>     TableClass = type(table_dict['class_name'], (Base,), table_dict)
> ```
> 
> An entry in `table_dict` will now look like:
> 
> ```
> {
>     'tablename': 'terms',
>     'columns': [
>         {'new_col': 'owner', 'type': ['UUID(as_uuid=True), default=uuid.uuid4, unique=True', 'nullable=False', 'index=True']},
>         {'new_col': 'received', 'type': ['Boolean', 'nullable=False']},
>         {'new_col': 'date_received', 'type': ['DateTime(timezone=True)', 'nullable=False']},
>     '__tablename__': 'terms',
>     'class_name': 'Terms'
>     ]
> }
> ```

I'm not sure if you are paraphrasing above, but the `table_dict` as written would not be passed directly to `type()`.  See my example which illustrates creating a new dictionary that features the ``__tablename__`` attribute as well as the assignment of ``Column()`` objects to attribute names on the class directly, extracting them from the original "columns" list that's in our arbitrary data structure.

> 
> Now we are at step 3! We want to commit this empty table, with all its columns and constraints, to the database.

I think you are referring to not what we usually call "commit this table" but rather to emit CREATE TABLE statements for the table to the database.  While some database backends do in fact support what we call "transactional DDL" that waits for a call to COMMIT to ensure the tables are created, that's not really the main thing we are doing here, the main thing you'd be looking to do is to emit the DDL for CREATE TABLE.

We do this by running either ``Base.metadata.create_all(e)`` from your Base class, or per table you can do `MyClass.__table__.create(e)`.



> 
> So after we call our function and have our object `TableClass` we start the engine:
> 
> ```
> Base.metadata.create_all(bind=engine)
> with db_session() as sess:
>      sess.add_all(TableClass(**rec) for rec in TableClass)
>      sess.commit()
> ```
> 
> This errors with `TypeError: 'DeclarativeMeta' object is not iterable` which makes sense.

this is likely due to the way you are creating the type() where as mentioned above you need to pass the third argment, the class dictionary, in the format that a declarative class would use.  Try running my example, it works as given, and then see what's different about your example vs. the one I provided.


> 
> If I pass in the dict as a whole instead of the TableClass object it errors `TypeError: DeclarativeMeta object argument after ** must be a mapping, not str` This makes sense given that we've ignored the TableClass object and are just dealing with a dict.

yes i think your class creation routine is not working.   I would be able to tell you more about that error message if you provided a complete stack trace but please try working with the code example I provided to understand the mechanics.

> It seems like I'm still making the same error as in my initial question! 

I agree!

> Have you any thoughts on how to read the table dictionary correctly into the Class correctly into the database? Thank you!!

yes, I provided you a complete, runnable example above.  Please work with that, thanks!

**Reply by fleetofthought** [reply]
@zzzeek Thank you for your thorough responses, really appreciate your patience and time!

I have reworked my code and have found a clearer core issue in how I'm trying to dynamically create a Table's Columns.

Here is an example of what the json looks like:

```
  "groupings": {
    "imaging": {
      "owner": { "type": "uuid", "required": true, "index": true },
      "tags": { "type": "text", "default": "#00C7FF"},
      "filename": { "type": "text" },
    },

    "user": {
      "email": { "type": "text", "required": true, "unique": true },
      "name": { "type": "text" },
      "role": {
        "type": "text",
        "required": true,
        "values": [
          "admin",
          "customer",
        ],
        "index": true
      },
      "date_last_logged": { "type": "timestamptz" }
    }
  }
```

When I parse the json to derive the values for each `Column`, there are different if conditions it passes through before being added to `json_cls_schema['columns']` because some of the values might be an actual type (Integer) while others might just need a constraint like (`'default='admin'`, or `'unique=True'`). After each condition the derived value gets added to a list and once all the values for that `Column` have been evaluated, that list is what gets updated into into `clsdict`.

```
from sqlalchemy import BigInteger, Boolean, Column, Constraint, DateTime, Integer, Table, Text
from sqlalchemy.dialects.postgresql import UUID, DOUBLE_PRECISION
from sqlalchemy.orm import declarative_base
from sqlalchemy import create_engine
import json

Base = declarative_base()

_type_lookup = {
    'bigint': BigInteger,
    'boolean': Boolean,
    'bool': Boolean,
    'double precision': DOUBLE_PRECISION,
    'index': 'index=True',
    'int': Integer,
    'required': 'nullable=False',
    'text': Text,
    'text[]': Text,
    'timestamptz': DateTime(timezone=True),
    'uuid': UUID,
    'unique': 'unique=True',
    'values': Text
}

def convert_table_classname(name):
    return ''.join(x.capitalize() or '_' for x in name.split('_'))

with open('data.json') as f:
    schema = json.load(f)

    # Select correct dict
    if schema_key == 'groupings':

    for schema_key, schema_value in schema.items():
        for table_name, col_data in schema_value.items():
            json_cls_schema = {'clsname': convert_table_classname(table_name),
                               'tablename': table_name,
                               'columns': [{"name": "col1", "type": Integer, "is_pk": True},]}

            for col_name, constraints in col_data.items():

                # Create list for the parsed values waiting to be inserted into table
                updated_value = []

                # 'default' doesn't need a type look up, just derive its value and render it into format required by Column
                if 'default' in constraints:
                    updated_value.append(f'default={constraints["default"]}')

                if 'type' in constraints:
                    if constraints['type'] in _type_lookup:
                        updated_value.append(_type_lookup[constraints['type']])

                # Requires extra step for nested values
                for x in constraints:
                    if x in _type_lookup:
                        updated_value.append(_type_lookup[x])
                # Append all the parsed values into json_cls_schema
                json_cls_schema['columns'].append({'name': col_name, 'type': updated_value})
```

Now we have all the column names and constraints. `json_cls_schema` looks like this:

```
{'clsname': 'Imaging',
 'tablename': 'imaging',
    'columns': [{'name': 'col1', 'type': <class 'sqlalchemy.sql.sqltypes.Integer'>, 'is_pk': True},
                {'name': 'owner', 'type': [<class 'sqlalchemy.dialects.postgresql.base.UUID'>, 'nullable=False', 'index=True']},
                {'name': 'tags', 'type': [<class 'sqlalchemy.sql.sqltypes.Text'>, 'default=#00C7FF']},
                {'name': 'filename', 'type': [<class 'sqlalchemy.sql.sqltypes.Text'>]}
                ]}

{'clsname': 'User',
 'tablename': 'user',
    'columns': [{'name': 'col1', 'type': <class 'sqlalchemy.sql.sqltypes.Integer'>, 'is_pk': True},
    {'name': 'email', 'type': [<class 'sqlalchemy.sql.sqltypes.Text'>, 'nullable=False', 'unique=True']},
    {'name': 'name', 'type': [<class 'sqlalchemy.sql.sqltypes.Text'>]},
    {'name': 'role', 'type': [<class 'sqlalchemy.sql.sqltypes.Text'>, 'nullable=False', <class 'sqlalchemy.sql.sqltypes.Text'>, 'index=True']},
    {'name': 'org', 'type': [<class 'sqlalchemy.sql.sqltypes.Text'>, 'nullable=False', 'index=True']},
    {'name': 'date_last_logged', 'type': [DateTime(timezone=True)]}]}
```

If the table class Imaging were hardcoded it should look like:
```
Class Imaging(Base):
    __tablename__ = 'imaging'
    col1 = Column(Integer, primary_key=True)
    owner = Column(UUID(as_uuid=True), default=uuid.uuid4, unique=True, index=True)
    tags = Column(Text, default='#00C7FF')
    filename = Column(Text)

```

The challenge with using the list is that when we try to do something like this:
```
clsdict = {"__tablename__": json_cls_schema["tablename"]}

    clsdict.update(
        {record["name"]: Column(
            record["type"], primary_key=record.get("is_pk", False)
        )
            for record in json_cls_schema["columns"]
        }
    )

    MyClass = type(json_cls_schema["clsname"], (Base,), clsdict)
    e = create_engine("sqlite://", echo=True)
    Base.metadata.create_all(e)
```

There are various errors relating to trying to pass a list, which all make sense!
`AttributeError: 'list' object has no attribute '_set_parent_with_dispatch'`
```
 sqlalchemy.exc.ArgumentError: 'SchemaItem' object, such as a 'Column' or a 'Constraint' expected,
got [<class 'sqlalchemy.sql.sqltypes.BigInteger'>, 'nullable=False']
```

So I tried wrapping the values in a Constraint at different points in the code. ex:
```
if 'default' in constraints:
    updated_value.append(Constraint(f'default={constraints["default"]}')) 
```
Or
```
json_cls_schema['columns'].append(Constraint({'name': column_name, 'type': updated_value})) 

TypeError: 'Constraint' object is not subscriptable
```
```
json_cls_schema['columns'].append({'name': column_name, 'type': Constraint(updated_value)})  

sqlalchemy.exc.CompileError: (in table 'x', column 'y'): Can't generate DDL for NullType(); did you forget to specify a type on this Column?
```
All of these still end up in a list since they're being appended

```
clsdict.update(
            {record['name']: Column(
                record['type'], Constraint(primary_key=record.get('is_pk', False))
            )
                for record in json_cls_schema["columns"]
            }
        )
sqlalchemy.exc.ArgumentError: 'SchemaItem' object, such as a 'Column' or a 'Constraint' expected, got [<class 'sqlalchemy.sql.sqltypes.BigInteger'>, 'nullable=False']
```

My approach to collecting the column constraints and then unpacking them is not working correctly due to appending values to the list `['columns']` in dict `json_cls_schema`. What would be a saner approach to run a column's values through multiple conditions and gather them before updating them directly into `clssdict'?

**Reply by zzzeek** [reply]
can you share the complete function that is working with clsdict?  the fragment as given looks fine and I dont see it creating a "list".   the only thing I can think of is that the symbol ``Column`` is not pointing to ``sqlalchemy.Column`` and instead some other object.  

this is why if you can provide a **single, succinct** .py file that includes *everything* to reproduce the bug I can just fix your scripts. with only fragments and no stack traces I can only guess.    editing the example I provided originally to reproduce your problem would be your best bet since you'd find the mistake on your own, i.e. change one line at a time, re-run, keep making it look like the non-working code until you figure out what's different.

**Reply by CaselIT** [reply]
The issue is that you are passing `'type': [<class 'sqlalchemy.sql.sqltypes.Text'>, 'nullable=False', 'index=True']` to column.
This equivalent to calling `Column([<class 'sqlalchemy.sql.sqltypes.Text'>, 'nullable=False', 'index=True']`. You cannot do that.

Also please note that `'nullable=False'` is a string not a python command, so if you pass it to the class you are calling it as `Column('nullable=False')` not as `Column(nullable=False)`. These are not the same.

**Reply by zzzeek** [reply]
oh yes in the data "type" is a list, indeed, thought that was a type object.  puzzle solved

**Comment by fleetofthought** [comment, marked as answer]
@zzzeek @CaselIT Thank you for all your help! 
Yes, the data type I was trying to pass was incorrect, and I had a couple of other errors as well in the way I was trying to gather all the columns and their data before committing the table.

Including my working code below in case it helps anyone in the future.

```py
from sqlalchemy import BigInteger, Boolean, Column, DateTime, Integer, Text
from sqlalchemy import create_engine
from sqlalchemy.orm import declarative_base

Base = declarative_base()

_type_lookup = {
    'bigint': BigInteger,
    'boolean': Boolean,
    'default': '',
    'index': True,
    'int': Integer,
    'required': False,
    'text': Text,
    'timestamptz': DateTime(timezone=True),
    'unique': True,
    'values': Text
}


def convert_table_classname(name):
    return ''.join(x.capitalize() or '_' for x in name.split('_'))


def mapping_for_json(schema):
    for table_key, table_values in schema.items():

        for table_name, col_data in table_values.items():

            json_cls_schema = {'tablename': table_name,
                               'columns': [{'name': 'id', 'type': Integer, 'is_pk': True}]}

            for col_name, constraints in col_data.items():
                new_col = {'name': col_name}

                if 'type' in constraints and constraints['type'] in _type_lookup:
                    new_col.update({'name': col_name, 'type': _type_lookup[constraints['type']]})

                if 'default' in constraints:
                    _type_lookup.update({'default': constraints['default']})

                for x in constraints:
                    if x is not 'type' and x in _type_lookup:
                        new_col.update({'name': col_name, 'type': _type_lookup[constraints['type']], x: _type_lookup[x]})

                json_cls_schema['columns'].append(new_col)

            clsdict = {'clsname': convert_table_classname(table_name), '__tablename__': json_cls_schema['tablename']}

            clsdict.update(
                {record['name']: Column(
                    record['type'],
                    primary_key=record.get('is_pk', False),
                    nullable=record.get('required', False),
                    unique=record.get('unique', False),
                    default=record.get('default', False)
                )
                    for record in json_cls_schema['columns']
                }
            )

            e = create_engine("sqlite://", echo=True)
            type(clsdict['clsname'], (Base,), clsdict)
            Base.metadata.create_all(e)


if __name__ == "__main__":
    json_data = {"groupings": {
        "imaging": {
            "owner": {"type": "bigint", "required": True, "index": True}, 
            "tags": {"type": "text", "default": "#00C7FF"},
            "filename": {"type": "text"},
        },

        "user": {
            "email": {"type": "text", "required": True, "unique": True},
            "name": {"type": "text"},
            "role": {
                "type": "text",
                "required": True,
                "values": [
                    "admin",
                    "customer",
                ],
                "index": True
            },
            "date_last_logged": {"type": "timestamptz"}
        }
    }
    }

    mapping_for_json(json_data)

```

**Reply by CaselIT** [reply]
thanks for sharing your working solution. hopefully it will be useful to others
